### ISSUE TITLE
CFU EventLog gives wrong status.

### ISSUE THREAD

**Issue by lscheffler** [VFPX/Thor#150]
## 📝 Provide detailed reproduction steps (if any)

1. CFU Download fails 

### ✔️ Expected result

Eventlog shows information URL runs
Eventlog shows fail of download

### ❌ Actual result

Eventlog does not show the URL   
Eventlog shows no error, downloads the file with the error and runs this file. This will fail, because it's just an errorlog   
Eventlog does give no information like errorcode etc

## 📷 Debugging Info
not applicable

## 📷 examples
Log:
```
18:52:50  Requesting file FoxBin2PRGVersion.txt?=_6OR14GUJ8 from server.
						 thor_proc_downloadfilefromurl.prg, line 81
						 thor_proc_getavailableversioninfo.prg, line 19
						 thor_proc_downloadandinstallupdates.prg, line 88
						 thor_proc_downloadandinstallupdates.prg, line 40
						 thor_proc_downloadandinstallupdates.prg, line 6

18:52:50  Downloading...
						 thor_proc_downloadfilefromurl.prg, line 99
						 thor_proc_getavailableversioninfo.prg, line 19
						 thor_proc_downloadandinstallupdates.prg, line 88
						 thor_proc_downloadandinstallupdates.prg, line 40
						 thor_proc_downloadandinstallupdates.prg, line 6

18:52:50  Download complete.
						 thor_proc_downloadfilefromurl.prg, line 134
						 thor_proc_getavailableversioninfo.prg, line 19
						 thor_proc_downloadandinstallupdates.prg, line 88
						 thor_proc_downloadandinstallupdates.prg, line 40
						 thor_proc_downloadandinstallupdates.prg, line 6

18:52:50  Executing code from downloaded version file...
						 thor_proc_getavailableversioninfo.prg, line 81
						 thor_proc_getavailableversioninfo.prg, line 26
						 thor_proc_downloadandinstallupdates.prg, line 88
						 thor_proc_downloadandinstallupdates.prg, line 40
						 thor_proc_downloadandinstallupdates.prg, line 6
```
file downloaded:
```
404: Not Found
```

I can provide an example prg to test.

**Comment by Jimrnelson** [comment]
Yup, thor_proc_downloadfilefromurl.prg does not recognize when it's been given an invalid URL.

This is out of my wheelhouse to resolve.

Am hoping somebody else will be able to do so.

**Comment by lscheffler** [comment]
At least you could hint the resulting error MIGHT be because of faulty URL - and display the URL.

**Comment by lscheffler** [comment]
We do not need to argue why the error happens. (In this case, a string long fixed that git somehow produced in a merge.) Only that there must be a chance to find it in the log. Here, even the path to the file with the error would have solved the problem quick. I mean _error running blabal.prg_ means edit the file - and _404: Not found_ is not hard to understand. And having the URL visible is quick to check.

**Comment by lscheffler** [comment]
Ah, and it would be even greater to get errors printed to the screen - so it's a failing silent, and one has no chance if one does not expect something. With the location of the EventLog etc.

**Comment by Jimrnelson** [comment]
> We do not need to argue why the error happens. (In this case, a string long fixed that git somehow produced in a merge.) Only that there must be a chance to find it in the log. Here, even the path to the file with the error would have solved the problem quick. I mean _error running blabal.prg_ means edit the file - and _404: Not found_ is not hard to understand. And having the URL visible is quick to check.

We are not arguing about why the error happens.  We are identifying where the actual error occurs and how it should be handled **at that place**.  The proc in question is erroneously reporting that it downloaded a file when it did not.  That should be fixed.

Anything that happens thereafter due to the erroneous report is just a symptom.  

BTW, you said "CFU fails".  How were you aware of that?  (I assume not from an error screen within CFU, else you would have sent it)

**Comment by Jimrnelson** [comment]
> Ah, and it would be even greater to get errors printed to the screen - so it's a failing silent, and one has no chance if one does not expect something. With the location of the EventLog etc.

You never reported what error was encountered, so this is difficult to address.

**Comment by lscheffler** [comment]
Oh, it's a file with a wrong string in My Updates that results in a wrong URL (FoxBin2Prg updating from my fork). Teh testing broken zip. I just managed to put the file to it's old state.
- First and major. There is no error encountered. It just silent ignores the file. No message, no nothing. Only I knew it's there. This is why I came up with #142
- All report is in OM. This is ALL I found.
- The error itself is the same as the one I fixed with #142, only that I have learned since then to find the files in the temp folder and learned to look  up the EventLog

Sry, this is all knowledge I have.

**Comment by Jimrnelson** [comment]
@lscheffler @DougHennig 

A number of points:

[1]  It would have been very helpful if in your original message you had explained more about the problem than 'CFU download failed'.  You gave no hint as to what the indication of failure was.  If you had said "Ran CFU without any reported errors.  However, something was wrong because ...".  This would have given this thread a much different start.  Even now, I do not know what you saw (or didn't see) that told you there was something wrong.

[2]  You found and pasted in part of the CFU event log.  Next time, please attach the entire log, as the part you pasted in did not answer all my questions.  (BTW, the CFU folder name is saved in `_Screen.cThorCFUFolder`)

[3]  Yeah, a bad URL for the version file.  Oops, but CFU should report this.

[4]  If the URL is not in github.com, CFU actually does report the problem with a MessageBox that looks like this:

![image](https://github.com/VFPX/Thor/assets/81179335/aea1c6fa-ba6d-47b3-ac01-5a99c85e5c5c)

[5] However, the proc (`thor_proc_downloadfilefromurl.prg`) that handles this does not recognize 'file not found' if the URL is in github.com, so it goes it's merry way, thinking it received a file when actually it did not.

[6]  There was no indication in the CFU log of any issue with the downloaded version file (which failed as well, because it could not be executed.)

So, needed fixes:

[1]  Fixing `thor_proc_downloadfilefromurl.prg` to return .F. for file not found.  I am hoping either you or Doug can do this, as I have no familiarity with downloading from the internet.

[2]  Expanding the EventLog to record URLs and destination files when downloading.  I have already changed this, but it cannot be released until 1.46.01 is done.

[3]  There is code to report errors in the version file, but, for reasons no longer applicable, it was skipped.  I have not corrected this yet.

**Comment by Jimrnelson** [comment]
@lscheffler 

> 2.  Lot of bloat, since this all repeats endless and put errors on every try. If you need more then the snippet, why not ask, I can post later?

I would not know exactly what to ask for in the event log.  However, I  assume that if I were able to look at the whole file something might well jump out at me.

That's why I asked you to attach the file (not paste).  I understand you can just drag a file to attach it, which is what I just did here:  [EventLog.txt](https://github.com/VFPX/Thor/files/12352539/EventLog.txt)

> 3. The URL is wrong in the sense that the capitalisation of all files used in the update-file was wrong: FoxBin2P_RG_ instead of FoxBin2Prg. You can create this with any update-file, just copy one to My Updates and change capitalization. For all I remember there was even an issue about this special error of mine - somewhere.

It's not of much interest why the URL was wrong, CFU should be able to handle and report on an incorrect URL.  I certainly have noticed the need for precise capitalization.

> 1. Me has no idea about programming web stuff. Only the common basics - for example that there is a return code for a failure. And that this crap is picky after the domain part. All I see is that one must check the file downloaded - if it is a short one-liner and the first 3 bytes are out of the html errorcode numbers, there will be a failure.

Rats!  Well then, I will tackle this unless @DougHennig has some experience with it.  Your suggestion looks easy enough.

> 2. You can push as long as you like, this would just need to merge 1.46.01 later. Nothing fancy. Might be a bit odd due to the included binary sources, but nothing one can not solve with git checkout.

No, one thing at a time.  I will wait with these changes until Doug has processed your pull request.

**Comment by lscheffler** [comment]
Only I had identified the part of the log with the message (or I would have no idea where to look)  So in generall there was not much else to search. Only if the important part is not on the snippet. Ouch.
Independent how to store the log, it is bloat I tend to post only on request.  :)

See https://github.com/VFPX/Thor/pull/149#issuecomment-1680113560, this is not a problem, rather common use of git(hub). It's a problem to me, since I have a problem testing my projects, but this should not be your concern. Except that sooner or later I must fix the GF stuff.

**Comment by Jimrnelson** [comment]
> Only I had identified the part of the log with the message (or I would have no idea where to look) So in generall there was not much else to search. Only if the important part is not on the snippet. Ouch.

Let me get this straight.  

You are the one reporting the problem.

And you're telling me that you know better than me about whether I can find anything of value in the entire event log file?  You could not possibly know that.

> Independent how to store the log, it is bloat I tend to post only on request. :)

Why do you care?  It's just an attachment, why does the size or structure matter?

I am not saying that the snippet you pasted wasn't of value. Rather that I expected you to attach the entire file as well.

**Comment by lscheffler** [comment]
> And you're telling me that you know better than me about whether I can find anything of value in the entire event log file? You could not possibly know that.

At least I told how to solve it as well. ¯\\\_(ツ)\_/¯

**Comment by Jimrnelson** [comment]
I believe that this has been fixed in Thor 1.46.01, along with some improvements in the event log.

**Comment by lscheffler** [comment]
Looks good.
